### ISSUE TITLE
Why the hack are terminal emulators actually (that) slow?

### ISSUE THREAD

**Issue by jerch** [cmuratori/refterm#15]
Some disclaimer upfront - I am hacking on terminal emulators (TE) for quite some years and I am mainly focused on speed pretty much like you suggested with your video (but more on the parser side, not the screen output). I am not trying to justify anything about the "new shiny kid in town" (windows terminal), what ever they did right/wrong, I cant be bothered less. Still I feel the need to speak up here, since your video might be misleading for people and what they can or should expect from a super performant TE.

I will list here some runtime results from my very own experiments with an older sandy bridge i7 (2.4 GHz with 1333 MHz RAM), that might help to identify the real bottlenecks and might give a more realistic approximation, of what would be a possible overall throughput. Where not explicitly stated otherwise, the numbers are derived from test runs on Ubuntu 18 with typical terminal data as of `ls -lR /` (with an SSD holding the filesystem). Note that I have no clue about Windows and its ConPty abstraction (but know for sure, that this creates its own bottleneck for fast producers).

Main things that would alter TE throughput on POSIX:

- OS pipe speed
  I list this here to get an idea what the system would be capable of in terms of raw throughput. For the machine above thats ~6GB/s with a producer maxing out the system's PIPE_BUF limit with normal `write` syscalls (not using special mapping tricks or `writev`). This will highly vary on faster machines, esp. with higher mem bus speeds. Lets take this one as the upper limit. Feeding a TE over an OS pipe is somewhat similar to what you did with fasterpipe, even more "naked". But we cannot use that directly, since we all love atleast some of the TTY semantics.

- kernel interface (PTY/TTY)
  One of the oldest piece of code in most kernels that did not change much over the last 40ys is the TTY abstraction. In theory it is just a small extension to OS pipes to insert process control, a line editor and some other nifty flags and rewrites, but in reality it sucks hard on the data throughput. The canonical mode for the line editor alone will create at least one memory direction copying things over to its own line buffer, and doing several char-by-char evaluations and such. With the machine above throughput will already drop to ~1GB/s for raw mode, and **~200MB/s** in canonical mode. Since most cmdline apps dont bother much for termios settings, they run with the default, which is canonical mode. Thus we already take longer than your 2s for that 1GB of data, if we dont fix things in the kernels to allow some faster transport.

- VT parsing speed
 For peeps not used to inner terminal mechanics It might be important to note here that you only do a small part of VT parsing. Ofc TEs implement very different sets of VT functions in the end, most stick in the DEC VT102 - VT240 realms and some ECMA-48 extensions. Proper VT parsing is a big issue and most TEs dont get it right for all sequences/functions (including my own parsers) for various reasons (lets not bore peeps here with that). Well, the state transitions are cumbersome with several side paths to be considered creating lots of branching which in return creates a big dent in the throughput. TEs have different strategies to deal with that complexity, from char-by-char branching (older ones) to a big DFA derived from VT100.net's parser model (most newer ones).
I have some experience here trying to squeeze out high perf from a VT parser doing proper sequence handling. The sad part about it - my fastest utf8 based parser is a C monster with ugly macro definitions and computed gotos - and it cant go faster than **~400 MB/s** on its own (data directly fed to the parser). Ofc this might be me doing bad optimizations and getting asm output all wrong, but comparing to other TEs thats right at the upper end of the parsing ballpark. Seems the "ops per char pressure" is simply that high or cant be reduced much further.

So far I did not mention any own buffer book keeping (scrollback?) and rendering in the TE at all, and wont even get to it. Just lets add up these numbers to get a more realistic expectation, what a TE on its own could change here (thus not changing kernel stuff):

1GB of data at:
- pty/tty: 200 MB/s --> 5s
- vt parsing: 400 MB/s --> 2.5s

means ~133 MB/s of "input throughput". Lets be generous and give newer machines a perf bonus of 3 times (!), and we are roughly in the range of **100 - 400 MB/s**. Thats imho what a TE could get to if it highly optimizes on the input chain, without considering inner buffering and the rendering. Without touching kernel interface or getting a significantly faster CPU (single core!) a number in the GBs is def. not in the books because of VT parsing alone.

Btw rendering is indeed pretty cheap compared to optimizing the data pumping :smile_cat: 

**TL;DR**
Imho a TE is unlikely surpass **100 - 400 MB/s** on nowadays machines and the near future for the input chain due to the nature of sequential terminal data and the dependency on (slower) kernel interfaces. The latter might be a bigger issue on windows, yet also affects POSIX systems.

@cmuratori Feel free to delete this, if you think it does not suit the bigger picture/goal of your repo.

**Comment by cmuratori** [comment]
I think a fair bit of this is not accurate (or is no longer accurate, anyway).  Specifically:

* The kernel interface is not necessary anymore. Thankfully, Windows 10 introduced ConPTY redirection, so you can completely bypass conhost now and reimplement the compatibility layer yourself, which other projects are already working on (https://github.com/jfhs/handterm/).

* There is no way that VT parsing is limited to 400mb/s on a modern machine, and especially in the case _where there are no VT codes_. So while I both think your estimate is low (I suspect a 1gb/s single-threaded VT parser is not out of the question even when there are dense VT codes), I also am not sure why it's relevant, because 99% of the time the terminal _is not parsing_ VT codes, and merely scanning input to verify there are no VT codes is extremely inexpensive.

\- Casey

**Comment by jerch** [comment]
> ...
>     * There is no way that VT parsing is limited to 400mb/s on a modern machine, and especially in the case _where there are no VT codes_. So while I both think your estimate is low (I suspect a 1gb/s single-threaded VT parser is not out of the question even when there are dense VT codes), I also am not sure why it's relevant, because 99% of the time the terminal _is not parsing_ VT codes, and merely scanning input to verify there are no VT codes is extremely inexpensive.

Sure thing, plz prove me wrong with such a single threaded parser. Just a minor sidenote - UTF8 to codepoint conversion alone is in that ballpark you are claiming here, not to speak of grapheme clustering (which some TEs started to support).

**Comment by cmuratori** [comment]
You don't need to convert UTF8 to codepoints when you're streaming. You wait until you render. Grapheme clustering and codepoint conversion don't need to be done during streaming.

Anyway, I don't feel like this is an "issue" at all, so I am going to close this.  This seems more like a blog post than a refterm issue.

\- Casey

**Comment by jerch** [comment]
> You don't need to convert UTF8 to codepoints when you're streaming. You wait until you render. Grapheme clustering and codepoint conversion don't need to be done during streaming.

This doesnt help at all, because current terminal mechanics are defined to be driven by cursor advance derived from a "runwidth" of a codepoint/cluster (`wcwidth`). Postponing the conversion just moves that ugly task to the render stage (or a VT sequence needing proper cursor positioning), you still have to do it once for all PRINT data to stay within the correct rows/columns progression.

> Anyway, I don't feel like this is an "issue" at all, so I am going to close this. This seems more like a blog post than a refterm issue.

Ofc it is not a technical issue of refterm, but more a way too eager promise I want to make ppl aware of. Just do the cycle maths with modern CPUs' clock speed, then things get pretty obvious about your 1gb/s claim.

Anyway, thx for your time.
